File: web/rpc/arguments/backend.js
try {
   var autobahn = require('autobahn');
   var when = require('when');
} catch (e) {
   // When running in browser, AutobahnJS will
   // be included without a module system
   var when = autobahn.when;
}

// dynamic connection uri based on file location
var wsuri;
if (document.location.origin == "file://") {
   wsuri = "ws://127.0.0.1:8080/ws";

} else {
   wsuri = (document.location.protocol === "http:" ? "ws:" : "wss:") + "//" +
               document.location.host + "/ws";
}

var connection = new autobahn.Connection({
   url: wsuri,
   realm: 'example_realm'}
);

connection.onopen = function (session) {

   function ping() {
   }

   function add2(args) {
      return args[0] + args[1];
   }

   function stars(args, kwargs) {
      kwargs = kwargs || {};
      kwargs.nick = kwargs.nick || "somebody";
      kwargs.stars = kwargs.stars || 0;
      return kwargs.nick + " starred " + kwargs.stars + "x";
   }

   var _orders = [];
   for (var i = 0; i < 50; ++i) _orders.push(i);

   function orders(args, kwargs) {
      kwargs = kwargs || {};
      kwargs.limit = kwargs.limit || 5;
      return _orders.slice(0, kwargs.limit);
   }

   function arglen(args, kwargs) {
      args = args || [];
      kwargs = kwargs || {};
      return [args.length, Object.keys(kwargs).length];
   }

   var dl = [];

   dl.push(session.register('examples.rpc.arguments.ping', ping));
   dl.push(session.register('examples.rpc.arguments.add2', add2));
   dl.push(session.register('examples.rpc.arguments.stars', stars));
   dl.push(session.register('examples.rpc.arguments.orders', orders));
   dl.push(session.register('examples.rpc.arguments.arglen', arglen));

   when.all(dl).then(
      function () {
         console.log("All registered.");
      },
      function () {
         console.log("Registration failed!", arguments);
      });  
};

connection.open();

File: examples/rpc_protocol/arguments/backend.py
from os import environ
from twisted.internet.defer import inlineCallbacks

from autobahn.twisted.wamp import ApplicationSession, ApplicationRunner


class Component(ApplicationSession):
    """
    An application component providing procedures with different kinds
    of arguments.
    """

    @inlineCallbacks
    def onJoin(self, details):
        print("session attached")

        def ping():
            return

        def add2(a, b):
            return a + b

        def stars(nick="somebody", stars=0):
            return u"{} starred {}x".format(nick, stars)

        def orders(product, limit=5):
            return [u"Product {}".format(i) for i in range(50)][:limit]

        def arglen(*args, **kwargs):
            return [len(args), len(kwargs)]

        yield self.register(ping, u'examples.rpc.arguments.ping')
        yield self.register(add2, u'examples.rpc.arguments.add2')
        yield self.register(stars, u'examples.rpc.arguments.stars')
        yield self.register(orders, u'examples.rpc.arguments.orders')
        yield self.register(arglen, u'examples.rpc.arguments.arglen')
        print("Procedures registered; ready for frontend.")


if __name__ == '__main__':
    runner = ApplicationRunner(
        environ.get("AUTOBAHN_DEMO_ROUTER", u"ws://127.0.0.1:8080/ws"),
        u"example_realm",
    )
    runner.run(Component)